packages/action-listener-middleware/src/index.ts

```diff
@@ -432,33 +432,51 @@ export function createActionListenerMiddleware<
     }
 
     // Need to get this state _before_ the reducer processes the action
-    const originalState = api.getState()
-    const getOriginalState = () => originalState
+    let originalState: S | typeof INTERNAL_NIL_TOKEN = api.getState()
+
+    // `getOriginalState` can only be called synchronously.
+    // @see https://github.com/reduxjs/redux-toolkit/discussions/1648#discussioncomment-1932820
+    const getOriginalState = (): S => {
+      if (originalState === INTERNAL_NIL_TOKEN) {
```

Comment:
`undefined` should really be enough here. No reducer is ever allowed to return `undefined`, so `getState()` cannot be `undefined` either.

Reply:
Maybe I'm missing something but I can't find an explicit invariant here: https://github.com/reduxjs/redux/blob/master/src/createStore.ts#L238

Where I can find it?

Reply:
Strictly speaking, I don't think we have such an invariant for the _root_ state.

`combineReducers` does have that check for _slice_ reducers, though:

https://github.com/reduxjs/redux/blob/793947f9a33d841ce91eb15dcaf3884d8aeb43b2/src/combineReducers.ts#L70-L75

Reply:
I'm aware and I'm sorry that's a really nitpicky  but you technically can use reducer: `() => undefined` with redux-toolkit.

DEMO: https://codesandbox.io/s/redux-store-reducer-undefined-mtner?file=/index.html

This demo shows that it is possible to end up in a `getOriginalState() === undefined` even when using  `configureStore` and `createSlice`.

Reply:
Huh, wasn't aware that that is combineReducers only. Then this seems necessary, yeah. Out of interest: does it go a few bytes down if the token is in the same file?

Reply:
Heh, this is actually a very similar issue to what I ran into with Reselect 4.1.x :)  Users occasionally wanted to have selectors that _do_ return `undefined`, so I had to use a similar `NOT_FOUND_TOKEN` approach internally.

Reply:
> Huh, wasn't aware that that is combineReducers only. Then this seems necessary, yeah. Out of interest: does it go a few bytes down if the token is in the same file?

It is slightly smaller `1589` vs `1592`.

Odd, as far as I can see `index.modern.js` does not preserve the original `esm` modules.

Reply:
It will create a new scope for the new file, a new variable to contain exports and inject that into the "module scope" of the other file. So it's a bit more code. But yeah, those are common to all modules so it makes sense that it gzips down to only 3 more bytes.